components/core/src/crypto/keys/cache.rs

```diff
@@ -0,0 +1,515 @@
+
+    /// Write a key into the cache. If the key already exists,
+    /// and the content has the same hash value, nothing will be
+    /// done. If the file exists and it has *different* content, an
+    /// Error is returned.
+    // Note that this is potentially problematic for public origin
+    // encryption keys and public origin signing keys. Even though
+    // they have different header values within a file, they could
+    // have the same name and revision (and thus the same filename)
+    // if new encryption and signing pairs were made within the same
+    // second.
+    //
+    // That is, if you made encryption and signing key pairs for
+    // "my-origin" in the same second, the public keys of both would
+    // both try to be saved in the same file, which is not great.
+    //
+    // This is probably _highly_ unlikely to manifest outside of
+    // tests, particularly since only Builder will be generating
+    // origin encryption pairs, but it's worth mentioning.
+    //
+    // This could also cause problems trying to retrieve one of these
+    // keys from the cache, either by name or named revision. If
+    // you're trying to fetch a public origin signing key, but happen
+    // to have a public origin *encryption* key in your cache, we'll
+    // find the file, but won't be able to give you a key, because it
+    // won't parse to the correct type.
+    //
+    // TODO (CM): Consider new filename extensions that capture what
+    // kind of key we're dealing with. For example "service.pub" and
+    // "service.box.key" for service encryption keys, "sig.pub"
+    // and "sig.key" for signing keys, etc. This would, of course,
+    // have to be done in a backwards-compatible way for all the keys
+    // currently in existence.
+    pub fn write_key<K>(&self, key: &K) -> Result<()>
+        where K: KeyFile
+    {
+        let keyfile = self.path_in_cache(key);
+        let content = key.to_key_string();
+
+        if keyfile.is_file() {
+            let existing_hash = hash::hash_file(&keyfile)?;
+            let new_hash = hash::hash_string(&content);
+            if existing_hash != new_hash {
+                let msg = format!("Existing key file {} found but new version hash is different, \
+                                   failing to write new file over existing. (existing = {}, \
+                                   incoming = {})",
+                                  keyfile.display(),
+                                  existing_hash,
+                                  new_hash);
+                return Err(Error::CryptoError(msg));
+            }
+        } else {
+            // Technically speaking, this probably doesn't really need
+            // to be an atomic write process, since we just tested
+            // that the file doesn't currently exist. It does,
+            // however, bundle up writing with platform-independent
+            // permission setting, which is *super* convenient.
```

Comment:
In addition to atomicity, the `AtomicWriter` also tries to assure the durability of file writes by blocking until changes have been flushed to disk with a [cross-platform](https://github.com/habitat-sh/habitat/blob/ffbb89e8590f2e9d4a61da5092d25937e3920b6e/components/core/src/fs.rs#L812-L863) implementation. Thus, I think it is definitely a wise choice to utilize the `AtomicWriter` for this process 🥰